File: example1/src/main/java/com/example/grpc/methodtypes/unary/UnaryAsynchronousClient.java
package com.example.grpc.methodtypes.unary;

import com.example.grpc.EchoRequest;
import com.example.grpc.EchoResponse;
import com.example.grpc.EchoServiceGrpc;
import com.example.grpc.Loggers;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

public class UnaryAsynchronousClient {

    private static final Logger logger = Logger.getLogger(UnaryAsynchronousClient.class.getName());

    public static void main(String[] args) throws Exception {
        Loggers.init();

        var channel = ManagedChannelBuilder.forAddress("localhost", 50051).usePlaintext().build();
        var asyncStub = EchoServiceGrpc.newStub(channel);
        var request = EchoRequest.newBuilder().setMessage("world").build();

        var done = new CountDownLatch(1);
        asyncStub.unaryEcho(request, new StreamObserver<>() {
            @Override
            public void onNext(EchoResponse response) {
                logger.log(Level.INFO, "response: {0}", response.getMessage());
            }

            @Override
            public void onError(Throwable t) {
                logger.log(Level.WARNING, "error: {0}", Status.fromThrowable(t));
                done.countDown();
            }

            @Override
            public void onCompleted() {
                logger.info("completed");
                done.countDown();
            }
        });

        done.await();
        channel.shutdown().awaitTermination(30, TimeUnit.SECONDS);
    }
}

File: _kotlin/src/main/java/com/example/grpc/echo/unary/UnaryAsynchronousClient.kt
package com.example.grpc.echo.unary

import com.example.grpc.EchoRequest

object  /*TODO*/ UnaryAsynchronousClient {
    private val logger: java.util.logging.Logger =
        java.util.logging.Logger.getLogger(UnaryAsynchronousClient::class.java.getName())

    @kotlin.Throws(java.lang.Exception::class)
    @kotlin.jvm.JvmStatic
    fun main(args: kotlin.Array<kotlin.String>) {
        Loggers.init()

        val channel: ManagedChannel = ManagedChannelBuilder.forAddress("localhost", 50051).usePlaintext().build()

        val asyncStub: Unit /* TODO: class org.jetbrains.kotlin.nj2k.types.JKJavaNullPrimitiveType */? =
            EchoServiceGrpc.newStub(channel)
        val request: Unit /* TODO: class org.jetbrains.kotlin.nj2k.types.JKJavaNullPrimitiveType */? =
            EchoRequest.newBuilder().setMessage("world").build()

        val done: CountDownLatch = CountDownLatch(1)
        asyncStub.unaryEcho(request, object : StreamObserver<V?> {
            public override fun onNext(response: EchoResponse) {
                UnaryAsynchronousClient.logger.log(java.util.logging.Level.INFO, "next: {0}", response.getMessage())
            }

            override fun onError(t: Throwable) {
                UnaryAsynchronousClient.logger.log(
                    java.util.logging.Level.WARNING,
                    "error: {0}",
                    io.grpc.Status.fromThrowable(t)
                )
                done.countDown()
            }

            override fun onCompleted() {
                UnaryAsynchronousClient.logger.info("completed")
                done.countDown()
            }
        })

        done.await()
        channel.shutdown().awaitTermination(30, TimeUnit.SECONDS)
    }
}
